### PR TITLE
More spawn doc

### PR DESCRIPTION
**Pull request by Elv13**
Inspired by #1273, https://github.com/mindeunix/clientdb/issues/, https://www.reddit.com/r/awesomewm/comments/5icvga/sending_notification_after_entering_the_password/ and some IRC questions over the last few months.

### CODE REVIEW

File: lib/awful/spawn.lua [hunk 1]
```diff
@@ -6,6 +6,61 @@
...
+-- documentation. It isn't usually necessary to even know it exists, as it is all
+-- done automatically. However, if more control is required, the identifier
+-- can be specified in by an environment variable called `DESKTOP_STARTUP_ID`.
+-- For example, if a command such as:
+--
+--    DESKTOP_STARTUP_ID="something" my_command
+--
+-- Is executed in a terminal, it should (if the program correctly implements the
+-- protocol) results in `c.startup_id` to at least match `something`. In case it
+-- is already used, some implementations will happen some suffixes to the string.
```
Comment (psychon) [inline lib/awful/spawn.lua]: Huh? Any examples for this?

As far as I remember, the `DESKTOP_STARTUP_ID` should have a special format and contain a timestamp. If there is no timestamp, GTK adds one.
How could anything even check if the startup id is already used?

File: lib/awful/spawn.lua [hunk 2]
```diff
@@ -6,6 +6,24 @@
...
+-- * When the command is a shell alias
+--
+-- Note there a shell is **not** a terminal emulator. A terminal emulator is
+-- something like XTerm, Gnome-terminal or Konsole. A shell is something like
+-- `bash`, `ZSH`, `busybox sh` or `Debian ash`.
+--
+-- If you wish to open a process in a terminal window, check that yours support
+-- the common `-e` option. If it does, then something like this should work:
+--
+--    awful.spawn(terminal.." -e 'my command'")
```
Comment (psychon) [inline lib/awful/spawn.lua]: Sadly, this is a problem.
https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=600429

With urxvt, you need the quotes. With gnome-terminal and terminator the quotes may not be present. Thus, you cannot portably execute a command containing spaces in a terminal.

I think this means we cannot describe this in the docs.

File: lib/awful/spawn.lua [hunk 3]
```diff
@@ -79,6 +79,52 @@
...
+-- First, do **not** use `io.popen` **ever**. It is synchronous. Synchronous
+-- function **stop everything** until they are done. All visual applications
+-- locks (as Awesome no longer respond), you will probably lose some keyboard
+-- and mouse events and will have higher latency when playing games. This is
+-- also true about reading files, but this is another topic.
+--
+-- Awesome provides a few ways of getting command results. One is to use the
+-- `Gio` libraries directly. This is usually very complicated, but gives a lot
+-- on control over the command execution, including the ability to create
+-- custom pipes and link many process.
```
Comment (psychon) [inline lib/awful/spawn.lua]: Uhm, I'm not sure if `Gio` really provides this, and if it does, then I bet that this is not accessible through `lgi`. Could we remove this part? (And just say that using `Gio` directly is complicated)

**Comment by psychon** [comment]
I added some fixup commits. If you agree with them, `git pull` to get them locally and `git rebase -i --autosquash master` to apply them. The comments that I left above are still open, I think (even though the GitHub interface disagrees).

File: lib/awful/spawn.lua [hunk 4]
```diff
@@ -130,20 +130,19 @@
...
--- to use the standard command. How the default application is selected is
+-- If the intent is to open a document, then it is recommended to use the
+-- following standard command. How the default application is selected is
 -- defined in the [Shared MIME-info Database](https://specifications.freedesktop.org/shared-mime-info-spec/shared-mime-info-spec-latest.html)
 -- specification. The `xdg-utils` package provided by most distributions
--- will provide the `xdg-open` command:
+-- provides the `xdg-open` command:
 --
---    awful.spawn("xdg-open /path/to/file")
+--    awful.spawn({"xdg-open", "/path/to/file"})
```
Comment (psychon) [inline lib/awful/spawn.lua]: Is this change ok?

I know that everyone thinks of commands as strings. However, the kernel (the `execve` syscall) wants an array of strings instead. Directly giving such an array to awesome avoids all kinds of word-splitting problems.

Sadly, this cannot really be used in the other examples...
(`noisy` could be turned into `noisy = { "bash", "-c", [[lots of text]] }`, but perhaps in this case that is more confusing than helping)

File: lib/awful/spawn.lua [hunk 5]
```diff
@@ -79,6 +79,52 @@
 --        placement = awful.placement.bottom_right,
 --    })'
 --
+-- **Getting a command result**:
+--
+-- First, do **not** use `io.popen` **ever**. It is synchronous. Synchronous
+-- function **stop everything** until they are done. All visual applications
+-- locks (as Awesome no longer respond), you will probably lose some keyboard
+-- and mouse events and will have higher latency when playing games. This is
```
Comment (Elv13) [inline lib/awful/spawn.lua]: > (but some key-bindings and focus-changing-keybindings will have non-intuitive results).

Or not happen at at. I have this issue when I compile and move the mouse too fast (yes, some modules I use, like Vicious and Drawer, still use `popen`...). This is the kind of issues I am referring to. I found awesome to be unreliable when it is slowed down by blocking calls. My Awesome 3.4 config had serious issues with that.

**Comment by Elv13** [comment]
done

**Pull request merged**